Feature/bcp swap

jordaniza opened the pull request:
The swap modal for DEFI+S/L => DEFI++ is now extended to multiple assets, whilst requiring a single buy token.

This means we can swap:

- DEFI+L/DEFI+S -> DEFI++
- DEFI++/WETH/WBTC -> BCP

The swap modal is configured to add a flexible launcher and take a contract address. It can therefore be reused in several places as needed.

Alexintosh committed:
Basic UI

Alexintosh committed:
Merge branch 'development' of github.com:pie-dao/pools.piedao.org into feature/compunding-vedough

TotoFullStackDev committed:
completed compound action

TotoFullStackDev committed:
hardcoded 10 SLICE,to be removed before merge

TotoFullStackDev committed:
loading bug fixed

TotoFullStackDev committed:
added update staking data after compound

TotoFullStackDev committed:
updated text content

TotoFullStackDev committed:
adapted mobile responsiveness

TotoFullStackDev committed:
first skeleton for wkpi redeem

jordaniza committed:
added Stake modal for KPI options redemption

jordaniza committed:
added claim workflow

jordaniza committed:
amended page text

jordaniza committed:
removed harcoded balance

jordaniza committed:
removed hardcoded boolean

jordaniza committed:
added stake

jordaniza committed:
changed vedough stake messag

jordaniza committed:
update UI and timelock until 3rd april

jordaniza committed:
Merge branch 'development' of https://github.com/pie-dao/pools.piedao.org into development

jordaniza committed:
merge upstream

jordaniza committed:
merge_upstream

jordaniza committed:
merge upstream

jordaniza committed:
merge upstream

jordaniza committed:
working basic approval flow

jordaniza committed:
better validations

jordaniza committed:
merge upstream

jordaniza committed:
merge upstream development and fix exchange liquidity

jordaniza committed:
reverse dirty commit history

jordaniza committed:
enable prod key

jordaniza committed:
fix loading tokens bug

jordaniza committed:
post tx hooks

jordaniza committed:
removed button

jordaniza committed:
replaced button with banner

jordaniza committed:
added to all pages

jordaniza committed:
fixed approval and timeout errors

jordaniza committed:
fix - lock on main exchange modal after timeout

mutantpunk committed:
Minor fixes

mutantpunk committed:
Merge

mutantpunk committed:
Fix closed tag

jordaniza committed:
allow swaps for defi+s and defi+l on all pages

jordaniza committed:
label min is now set to 0

jordaniza committed:
entry into BCP

jordaniza committed:
modal has now got flexible buy token and slots

jordaniza committed:
added mobile image for merge

jordaniza committed:
merge mobile changes

gbonumore commented on src/components/JoinSwapExternAmount.svelte:
```diff
@@ -58,10 +57,10 @@
   $: tokensLoaded = tokenList.length > 0;
-  
-  
+  $: sellTokenDecimals = sellToken?.decimals ?? 18;
```
I think this could be included on the defaultAmount so you don't actually extend the default value for decimals here

jordaniza replied:
That's a very nice spot

gbonumore commented on src/components/JoinSwapExternAmount.svelte:
```diff
@@ -537,7 +547,7 @@
         modal.open();
       }}
-      disabled={error || isLoading || balanceError || amount.label <= 0}
+      disabled={error || isLoading || balanceError || sellAmount.label <= 0}
```
sellAmount.label cannot be < 0 from line 126, so it should be only sellAmount.label === 0

gbonumore commented on src/components/SingleAssetEntryExit.svelte:
```diff
@@ -0,0 +1,32 @@
+        />
+    </slot>
+    <Modal title={'Swap Tokens'} backgroundColor="#f3f3f3" bind:this={swapModal}>
+        <span class="w-full" slot="content">
```
Do not wrap an entire component with divs inside spans (spans should only allow inline elements not blocks)

jordaniza replied:
In the svelte docs they use spans: https://svelte.dev/tutorial/named-slots

jordaniza committed:
added code review changes

gbonumore commented on src/components/SingleAssetEntryExit.svelte:
```diff
@@ -0,0 +1,39 @@
+    <slot>
+        <img src={images.defi_pp_merge} alt="defi single asset exchange"
+            class="hidden md:block cursor-pointer"
+            on:click={() => { openTrigger() }}
```
I am a fan of just passing the references: on:click={openTrigger} but this is a matter of personal preference

jordaniza replied:
Certainly looks better

jordaniza committed:
trying joinswap

jordaniza committed:
Merge branch 'development' of https://github.com/pie-dao/pools.piedao.org into development

jordaniza committed:
exit possible with ui fixes needed

jordaniza committed:
ui for exit swaps

jordaniza committed:
exit swap works

jordaniza committed:
working modal for exit and entry

jordaniza committed:
fix conflicts and add conditional exit

jordaniza committed:
Merge branch 'development' into feature/bcp-swap

jordaniza committed:
auto buy/sell when selecting bcp

jordaniza committed:
styling changes

jordaniza committed:
improved loading

jordaniza commented:
Some additional changes:

- Fixed an issue whereby styling on large screens was broken
- Improved loading feedback
- Removed entry/exit and instead added auto detect exitPool if selecting the corresponding token
- Removed the dedicated button and conditionally added to the swap if you hit `swap` from the BCP page.

Outstanding issues:
- The currently design of the swap modal does not pre-fetch tokens, there is therefore a loading period while tokens are being fetched. Made some changes to improve loading speed and UI responsiveness but does not seem to be a quick way around that while keeping the token list dynamic.

jordaniza committed:
replace exit swap call and improve ui when not signed in

jordaniza commented:
@Alexintosh I updated the PR based on our discussion:

- When not signed in, we still populate the tokens list, ignoring balances
- Use the `exitswapPoolAmountIn` contract call instead of the other one

jordaniza committed:
added bcp when not signed in

jordaniza committed:
attach maxallowance on approvemax, add lifecycle hook for txpending

jordaniza committed:
Merge branch 'development' into feature/bcp-swap

jordaniza commented:
Based on comments I have added two further changes: 

1. Once the approveMax function has been called, we update the `sellToken` allowance:

```js
      await approveMax(sellToken.address, spender);
      sellToken.allowance = ethers.constants.MaxUint256;
```
 This prevents the issue whereby the user is waiting for approvals to update, straight after approving the contract.

2. Add a transaction lifecycle hook `isConfirmingTx` which will cause the swap button to be disabled and render `Swap in Progress...` until the swap resolves. This at the very least gives visual feedback to the user that the swap is happening and prevents accidentally wasting gas on a double spend

jordaniza committed:
re-add smart contracts to pieindexlanding
